### ISSUE TITLE
Cannot get taxonomizr to work

### ISSUE THREAD

**Issue by RBuesink** [sherrillmix/taxonomizr#60]
I tried to setup taxonomizr automatically (didn't work), although the files were downloaded. then manually tried, but at this step, it didn't work anymore. My code:

```
install.packages("taxonomizr")
library(taxonomizr)

setwd("C:/Users/rickb/OneDrive/Documents/taxonomizr")
getNamesAndNodes()
getAccession2taxid(baseUrl = "https://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/nucl_gb.accession2taxid.gz")
read.names.sql('names.dmp','accessionTaxa.sql')
read.nodes.sql('nodes.dmp','accessionTaxa.sql')
read.accession2taxid(list.files('.','accession2taxid.gz$'),'accessionTaxa.sql')

getAccession2taxid(baseUrl='https://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/')
prepareDatabase('accessionTaxa.sql')
getNamesAndNodes()
```

But what I get in return:

```

> setwd("C:/Users/rickb/OneDrive/taxonomizr")
Error in setwd("C:/Users/rickb/OneDrive/taxonomizr") : 
  cannot change working directory
> setwd("C:/Users/rickb/OneDrive/Documents/taxonomizr")
> getNamesAndNodes()
Download status: 1 done; 0 in progress. Total size: 58.02 Mb (100%)... done!             
[1] "./names.dmp" "./nodes.dmp"
> getAccession2taxid()
This can be a big (several gigabytes) download. Please be patient and use a fast connection.
Download status: 1 done; 0 in progress. Total size: 1988.39 Mb (100%)... done!             
Error in resumableDownload(xx, yy) : 
  Download failed with error: "transfer closed with 183148808 bytes remaining to read". Progress is saved in ./nucl_gb.accession2taxid.gz.__TMP__ and continued download can be attempted by repeating the previous command.
Delete ./nucl_gb.accession2taxid.gz.__TMP__ or set resume=FALSE to start from scratch.
> getAccession2taxid(resume = TRUE)
Error in getAccession2taxid(resume = TRUE) : 
  unused argument (resume = TRUE)
> getAccession2taxid()
This can be a big (several gigabytes) download. Please be patient and use a fast connection.
Download status: 1 done; 0 in progress. Total size: 2163.05 Mb (100%)... done!             
Error in resumableDownload(xx, yy) : 
  Download failed with error: "server did not report OK, got 450". Progress is saved in ./nucl_gb.accession2taxid.gz.__TMP__ and continued download can be attempted by repeating the previous command.
Delete ./nucl_gb.accession2taxid.gz.__TMP__ or set resume=FALSE to start from scratch.
> getAccession2taxid()
This can be a big (several gigabytes) download. Please be patient and use a fast connection.
Download status: 1 done; 0 in progress. Total size: 2163.05 Mb (100%)... done!             
 [100%] Downloaded 61 bytes...
Error in (function (xx, yy)  : 
  Downloaded file does not match ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/nucl_gb.accession2taxid.gz File corrupted or download ended early?
> getAccession2taxid(baseUrl = "ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/nucl_gb.accession2taxid.gz")
This can be a big (several gigabytes) download. Please be patient and use a fast connection.
Download status: 1 done; 0 in progress. Total size: 0 b ... done!             
Error in resumableDownload(xx, yy) : 
  Download failed with error: "Server denied you to change to the given directory".
> getAccession2taxid(baseUrl = "https://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/nucl_gb.accession2taxid.gz")
This can be a big (several gigabytes) download. Please be patient and use a fast connection.
Download status: 1 done; 0 in progress. Total size: 988 b (100%)... done!             
Download status: 1 done; 0 in progress. Total size: 988 b (100%)... done!             
[1] "./nucl_gb.accession2taxid.gz"  "./nucl_wgs.accession2taxid.gz"
> read.names.sql('names.dmp','accessionTaxa.sql')
> read.nodes.sql('nodes.dmp','accessionTaxa.sql')
> read.accession2taxid(list.files('.','accession2taxid.gz$'),'accessionTaxa.sql')
Reading nucl_gb.accession2taxid.gz.
Error in trimTaxa(ii, tmp, 1:3) : Malformed line on line 2 
>
```

I've been trying for days now (downloading each time takes long. Can you help me?

**Comment by sherrillmix** [comment]
It's a bit hard to parse your example (e.g. repeating the same step multiple times in odd orders) but it would seem that there's a bit of a disconnect between you and NCBI during the download of the big accession2taxid files. For example, it would seem that the `nucl_gb.accession2taxid.gz` file you downloaded is garbled in the first couple lines. You could double check this with something like `zcat nucl_gb.accession2taxid.gz|head` or in R `readLines('nucl_gb.accession2taxid.gz',n=10)`. When I do that I get:

```
> readLines('nucl_gb.accession2taxid.gz',n=10)
 [1] "accession\taccession.version\ttaxid\tgi"
 [2] "A00001\tA00001.1\t10641\t58418"         
 [3] "A00002\tA00002.1\t9913\t2"              
 [4] "A00003\tA00003.1\t9913\t3"              
 [5] "A00004\tA00004.1\t32630\t57971"         
 [6] "A00005\tA00005.1\t32630\t57972"         
 [7] "A00006\tA00006.1\t32630\t57973"         
 [8] "A00008\tA00008.1\t32630\t57974"         
 [9] "A00009\tA00009.1\t32630\t57975"         
[10] "A00010\tA00010.1\t32630\t57976"
```

but from that error, it seems likely something odd has happened in your file. 

Oh actually `getAccession2taxid(baseUrl = "https://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/nucl_gb.accession2taxid.gz")` is not correct since the `baseUrl` is supposed to point to the directory the files are located in i.e. `https://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/`. I do see that when one passes in an erroneous argument like that to the function, it still returns an empty file and no error. I need to catch that and generate an error. I guess that's what generating your final error here, an empty file from the prior erroneous command. 

I don't think that's the root cause of your trouble but it's pretty hard to debug with the overlapping problems. Perhaps try a simple `prepareDatabase('accessionTaxa.sql')` and if the download fails midway through (implying a hiccup either at NCBI or your internet), repeat the command until the file completely downloads (it will try to resume at the place where you disconnected rather than starting over) and then see if there is still an error.

**Comment by RBuesink** [comment]
Thanks for your reply,

I found the baseUrl-option somewhere online, but this apparently didn't
solve the problem.
The download after regular preparation indeed stopped after a while, but
re-starting it asked for deletion of the .sql file. After i did this, the
following:

> prepareDatabase('accessionTaxa.sql')SQLite database accessionTaxa.sql already exists. Delete to regenerate[1] "accessionTaxa.sql"> prepareDatabase('accessionTaxa.sql')Downloading names and nodes with getNamesAndNodes()
./names.dmp, ./nodes.dmp already exist. Delete to redownload
Preprocessing names with read.names.sql()Preprocessing nodes with
read.nodes.sql()Downloading accession2taxid with getAccession2taxid()
This can be a big (several gigabytes) download. Please be patient and
use a fast connection.Download status: 1 done; 0 in progress. Total
size: 2163.05 Mb (100%)... done!              [100%] Downloaded 61
bytes...Error in (function (xx, yy)  :
  Downloaded file does not match
ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/nucl_gb.accession2taxid.gz
File corrupted or download ended early?


I don't know how to solve this; manual download doesn't help either
(similar errors)

Op vr 14 apr 2023 om 20:38 schreef Scott Sherrill-Mix <
***@***.***>:

**Comment by sherrillmix** [comment]
Sounds like something is corrupting the download. Perhaps some sort of problem with internet? What do you mean by "manual download"? If you mean you tried to download manually outside R and failed then I guess that's pointing to a larger issue that I'm not sure I'll have a lot of insight on.

**Comment by sherrillmix** [comment]
I'm going to close but feel free to reopen or start new issue if there's still a problem.
